core/src/main/java/org/apache/calcite/rel/rules/SortRemoveRedundantRule.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.core.Sort;
+
+import org.immutables.value.Value;
+
+/**
+ * Planner rule that removes
+ * the redundant {@link org.apache.calcite.rel.core.Sort} if its input
+ * max row number is less than or equal to one.
+ *
+ * <p> For example:
+ * <blockquote><pre>{@code
+ *  select max(totalprice) from orders order by 1}
+ *  </pre></blockquote>
+ *
+ * <p> could be converted to
+ * <blockquote><pre>{@code
+ *  select max(totalprice) from orders}
+ *  </pre></blockquote>
+ *
+ * @see CoreRules#SORT_REMOVE_REDUNDANT
+ */
+@Value.Enclosing
+public class SortRemoveRedundantRule
+    extends RelRule<SortRemoveRedundantRule.Config>
+    implements TransformationRule {
+  protected SortRemoveRedundantRule(final SortRemoveRedundantRule.Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(final RelOptRuleCall call) {
+    final Sort sort = call.rel(0);
+    if (sort.offset != null || sort.fetch != null) {
```

Comment:
Could we also support explicit OFFSET and LIMIT when sort's input max row count is less or equal than 1? I think it is also useful.

Reply:
> Could we also support explicit OFFSET and LIMIT when sort's input max row count is less or equal than 1? I think it is also useful.

Hi @herunkang2018 ,thank you for your review suggestions.

We could remove the Limit if input's row number less or equal to Limit fetch ,and could also redundant topN when input row number less or equal to 1.

I plan to contribute these optimizations in the future.